### PR TITLE
Refactor the gRPC enumerators

### PR DESCRIPTION
**Pull request by dhingrak**
Added: Refactor gRPC enumerators 

Fixes https://linear.app/eventstore/issue/DB-392/refactor-the-grpc-enumerators

Decouple all the enumerators from gRPC so they can work directly with connectors.

### CODE REVIEW

File: src/EventStore.Core/Services/Transport/Grpc/Enumerators.AllSubscription.cs [hunk 1]
```diff
@@ -44,26 +42,16 @@ public class AllSubscription : IAsyncEnumerator<ReadResp> {
...
-				}
-
-				if (readIndex == null) {
-					throw new ArgumentNullException(nameof(readIndex));
-				}
-
 				_expiryStrategy = expiryStrategy;
 				_subscriptionId = Guid.NewGuid();
-				_bus = bus;
+				_bus = bus ?? throw new ArgumentNullException(nameof(bus));
```
Comment (timothycoleman) [inline src/EventStore.Core/Services/Transport/Grpc/Enumerators.AllSubscription.cs:48]: i could be persuaded, but i think i'd be tempted to leave these tidyups for another PR, so that this one can be as minimal and obviously correct as possible

File: src/EventStore.Core/Services/Transport/Grpc/Enumerators.AllSubscription.cs [hunk 2]
```diff
@@ -384,28 +390,15 @@ await foreach (var @event in liveEvents.Reader.ReadAllAsync(_cancellationToken)
...
 				? new ValueTask(Task.CompletedTask)
-				: _channel.Writer.WriteAsync(new ReadResp {
-					Confirmation = new ReadResp.Types.SubscriptionConfirmation {
-						SubscriptionId = SubscriptionId
-					}
-				}, _cancellationToken);
-
-			private void Fail(Exception exception) {
-				Interlocked.Exchange(ref _subscriptionStarted, 1);
-				_channel.Writer.TryComplete(exception);
```
Comment (timothycoleman) [inline src/EventStore.Core/Services/Transport/Grpc/Enumerators.AllSubscription.cs:407]: this closes the channel with an exception but most of that places that used to call this now close the channel cleanly or dont close it at all, is that an intentional change?

File: src/EventStore.Core/Services/Transport/Grpc/ReadResponse.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
+﻿using EventStore.Core.Data;
+
+namespace EventStore.Core.Services.Transport.Grpc; 
+
+public abstract class ReadResponse {
+	
+}
+
+public class ResolvedEventWrapper: ReadResponse {
```
Comment (timothycoleman) [inline src/EventStore.Core/Services/Transport/Grpc/ReadResponse.cs:9]: it might be nice if these were nested in the abstract base class, its a pattern we often do with the message classes

File: src/EventStore.Core/Services/Transport/Grpc/Enumerators.AllSubscription.cs (hunk 1 above)
Comment (dhingrak) [reply src/EventStore.Core/Services/Transport/Grpc/Enumerators.AllSubscription.cs:48]: sure I will update this.

File: src/EventStore.Core/Services/Transport/Grpc/ReadResponse.cs (hunk 3 above)
Comment (dhingrak) [reply src/EventStore.Core/Services/Transport/Grpc/ReadResponse.cs:9]: Okay

**Review by timothycoleman — APPROVED** [review]
thanks! :+1:

**Review by shaan1337 — APPROVED** [review]
